File: frontend/src/types/agent.ts
/**
 * TypeScript interfaces for agent chat functionality.
 */

export interface AgentStatus {
	available: boolean;
	provider: string;
	model: string | null;
	missing_config: string[];
	message: string;
}

export interface ChatSession {
	id: number;
	session_id: string;
	user_id: string | null;
	context_summary: string | null;
	is_active: boolean;
	created_at: string;
	updated_at: string;
}

export interface ToolCall {
	tool_name: string;
	args: Record<string, any>;
	id: string;
	status: "executing" | "executed" | "requires_auth" | "failed";
	result?: string;
	auth_scope?: string;
	error?: string;
}

export interface ChatMessage {
	id: number;
	session_id: string;
	role: "user" | "assistant" | "system";
	content: string;
	tool_calls: ToolCall[] | null;
	metadata: Record<string, any> | null;
	token_count: number | null;
	created_at: string;
}

export interface AgentEventSubscription {
	id: number;
	session_id: string;
	task_id: number;
	event_types: string[];
	is_active: boolean;
	created_at: string;
	expires_at: string | null;
}

// Request types
export interface ChatSessionCreateRequest {
	user_id?: string;
}

export interface ChatMessageCreateRequest {
	content: string;
}

export interface AgentEventSubscriptionCreateRequest {
	task_id: number;
	event_types?: string[];
}

// Session sync types
export interface MessageSync {
	role: string;
	content: string;
	created_at: string;
}

export interface SessionSyncRequest {
	session_id: string;
	created_at: string;
	messages: MessageSync[];
}

export interface SessionListItem {
	session_id: string;
	created_at: string;
	updated_at: string;
	last_message_preview: string;
	message_count: number;
}

// Authorization types
export interface ToolAuthorizationRequest {
	scopes: string[];
	expires_at?: string;
}

export interface ToolExecutionResult {
	success: boolean;
	result: string;
	tool_name: string;
	call_id?: string;
	requires_auth?: boolean;
	auth_scope?: string;
	authorized?: boolean;
}

export interface AuthorizationResponse {
	status: "granted" | "revoked";
	scopes: string[];
	tool_results?: ToolExecutionResult[];  // Results of pending tool calls executed after authorization
}

// Unified iteration display type
export interface IterationBlock {
	iteration: number;          // 1-based iteration number
	content: string;            // Text content for this iteration
	toolCalls: ToolCall[];      // Tool calls WITH results attached
	status: 'streaming' | 'complete';
}

File: src/web/schemas/agent.py
"""
Pydantic schemas for agent API.
"""

from pydantic import BaseModel, Field
from datetime import datetime
from typing import Optional, List, Dict, Any


class ChatSessionCreate(BaseModel):
	"""Schema for creating a new chat session."""

	user_id: Optional[str] = None


class ChatSessionResponse(BaseModel):
	"""Schema for chat session response."""

	id: int
	session_id: str
	user_id: Optional[str]
	title: Optional[str]
	context_summary: Optional[str]
	is_active: bool
	created_at: datetime
	updated_at: datetime

	class Config:
		from_attributes = True


class ChatMessageCreate(BaseModel):
	"""Schema for sending a message."""

	content: str = Field(..., min_length=1, max_length=10000)


class ChatMessageResponse(BaseModel):
	"""Schema for chat message response."""

	id: int
	session_id: str
	role: str  # user, assistant, system
	content: str
	tool_calls: Optional[List[Dict[str, Any]]] = None
	message_metadata: Optional[Dict[str, Any]] = None
	token_count: Optional[int] = None
	created_at: datetime

	class Config:
		from_attributes = True


class AgentEventSubscriptionCreate(BaseModel):
	"""Schema for creating event subscription."""

	task_id: int
	event_types: List[str] = Field(default=["task_completed", "task_failed"])


class AgentEventSubscriptionResponse(BaseModel):
	"""Schema for event subscription response."""

	id: int
	session_id: str
	task_id: int
	event_types: List[str]
	is_active: bool
	created_at: datetime
	expires_at: Optional[datetime]

	class Config:
		from_attributes = True


class MessageSync(BaseModel):
	"""Schema for syncing a message from IndexedDB."""

	role: str
	content: str
	created_at: datetime


class SessionSyncRequest(BaseModel):
	"""Schema for syncing a full session from IndexedDB to backend."""

	session_id: str
	created_at: datetime
	messages: List[MessageSync]


class SessionListItem(BaseModel):
	"""Schema for session list item."""

	session_id: str
	created_at: datetime
	updated_at: datetime
	title: Optional[str]
	last_message_preview: str
	message_count: int


class TitleUpdateRequest(BaseModel):
	"""Schema for updating session title."""

	title: str = Field(..., min_length=1, max_length=100)
